packages/core/core/src/AssetGraphBuilder.js

```diff
@@ -164,11 +171,105 @@ export default class AssetGraphBuilder extends EventEmitter {
     }
   }
 
+  // Defer transforming this dependency if it is marked as weak, there are no side effects,
+  // no re-exported symbols are used by ancestor dependencies and the re-exporting asset isn't
+  // using a wildcard.
+  // This helps with performance building large libraries like `lodash-es`, which re-exports
+  // a huge number of functions since we can avoid even transforming the files that aren't used.
+  shouldDeferDependency(dependency: Dependency, sideEffects: ?boolean) {
+    let defer = false;
+    if (
+      dependency.isWeak &&
+      sideEffects === false &&
+      !dependency.symbols.has('*')
+    ) {
+      let depNode = this.assetGraph.getNode(dependency.id);
+      invariant(depNode);
+
+      let assets = this.assetGraph.getNodesConnectedTo(depNode);
+      let symbols = invertMap(dependency.symbols);
+      invariant(assets.length === 1);
+      let firstAsset = assets[0];
+      invariant(firstAsset.type === 'asset');
+      let resolvedAsset = firstAsset.value;
+      let deps = this.assetGraph.getIncomingDependencies(resolvedAsset);
+      defer = deps.every(
+        d =>
+          !d.symbols.has('*') &&
+          ![...d.symbols.keys()].some(symbol => {
+            let assetSymbol = resolvedAsset.symbols.get(symbol);
+            return assetSymbol != null && symbols.has(assetSymbol);
+          })
+      );
+    }
+    return defer;
+  }
+
   handleCompletedDepPathRequest(
     requestNode: DepPathRequestNode,
-    result: AssetRequest | null
+    assetGroup: AssetRequest | null
   ) {
-    this.assetGraph.resolveDependency(requestNode.value, result);
+    if (!assetGroup) {
+      return;
+    }
+    let dependency = requestNode.value;
+
+    let defer = this.shouldDeferDependency(dependency, assetGroup.sideEffects);
+
+    let assetGroupNode = nodeFromAssetGroup(assetGroup, defer);
+    let existingAssetGroupNode = this.assetGraph.getNode(assetGroupNode.id);
+    if (existingAssetGroupNode) {
+      // Don't overwrite non-deferred asset groups with deferred ones
+      invariant(existingAssetGroupNode.type === 'asset_group');
+      assetGroupNode.deferred = existingAssetGroupNode.deferred && defer;
+    }
+    this.assetGraph.resolveDependency(dependency, assetGroupNode);
+    if (existingAssetGroupNode) {
+      // Node already existed, that asset might have deferred dependencies,
+      // recheck all dependencies of all assets of this asset group
+      let assetNodes = this.assetGraph
```

Comment:
Maybe use `traverse` for this? You can use `skipChildren` to stop going any further once you hit the asset group.

Reply:
Great suggestion, that section is way more concise now.